onnxruntime/core/providers/cpu/tensor/upsample.cc

```diff
@@ -576,6 +577,58 @@ void UpsampleBilinear(const int64_t batch_size,
   }
 }
 
+template <typename T>
+void NhwcUpsampleBilinear(const int64_t batch_size,
+                          const int64_t num_channels,
+                          const int64_t input_height,
+                          const int64_t input_width,
+                          const int64_t output_height,
+                          const int64_t output_width,
+                          const float height_scale,
+                          const float width_scale,
+                          const std::vector<float>& roi,
+                          const bool use_extrapolation,
+                          const float extrapolation_value,
+                          const T* const XdataBase,
+                          T* const YdataBase,
+                          AllocatorPtr& alloc,
+                          const GetOriginalCoordinateFunc& get_original_coordinate,
+                          concurrency::ThreadPool* tp) {
+  BilinearParams p = SetupUpsampleBilinear(input_height, input_width, output_height, output_width,
+                                           height_scale, width_scale, roi,
+                                           alloc, get_original_coordinate);
+  for (int64_t n = 0; n < batch_size; ++n) {
+    concurrency::ThreadPool::TrySimpleParallelFor(
+        tp, num_channels,
```

Comment:
>num_channels [](http://example.com/codeflow?start=12&length=12)

from the perspective of performance, it doesn't make sense to parallel with channel any more because channel is the last dimension for NHWC. It is better to handle a channel count at a time.

Reply:
According to experiments results, parallel with output_height instead of num_channels is faster.

Reply:
Did you try parallel with output_height * output_width? What if output_height is very small but output_width is high

Reply:
Parallel with output_height * output_width instead.